File: pyround1/24.两两交换链表中的节点.py
#
# @lc app=leetcode.cn id=24 lang=python
#
# [24] 两两交换链表中的节点
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(-1)
        dummy.next = head
        pre = dummy
        while head and head.next:
            first = head
            second = head.next

            pre.next = second
            first.next = second.next
            second.next = first

            pre = first 
            head = first.next
        return dummy.next

# @lc code=end

File: 24.两两交换链表中的节点.cpp
/*
 * @lc app=leetcode.cn id=24 lang=cpp
 *
 * [24] 两两交换链表中的节点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if(!head)
            return nullptr;
        if(!head->next)
            return head;
        ListNode* pre = new ListNode(-1);
        pre->next = head;
        ListNode* res = pre;
        ListNode* fast = head->next;
        ListNode* slow = head;
        while(fast){
            pre->next = fast;
            slow->next = fast->next;
            fast->next = slow;
            pre = slow;
            slow = slow->next;
            if(!slow)
                break;
            fast = slow->next;
        }
        return res->next;

    }
};
// @lc code=end

